Fix nrepl port race condition

clyfe opened the pull request:
<!-- ❤️ Thanks for filing a Pull Request on Calva! You are contributing to a better Clojure coding experience. ❤️ -->
<!-- We use checklists in order to not forget about important lessons we and others have learnt along the way. -->

## What has Changed?
<!-- Introduce the change(s) briefly here. Consider explaining why a particular change was implemented the way it was. If you have considered alternative ways to introduce the change, please elaborate a bit on that as well. -->
- Race is: the get-port-from-file code overwrites the port-from-regex with nothing when the nrepl-port file is not yet present at the time is read.
- Fix: don't get-port-from-file when port already present.

<!-- Tell us what Github issue(s) your PR is fixing. Consider creating the issue if need be. -->
Fixes #901 

## My Calva PR Checklist
<!-- Remove the checkboxes that do not apply, as Github reports how many are not ticked. If you want to add checkboxes, please do. -->

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Tested the VSIX built **locally** from the PR (so, after you've submitted the PR).
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)

## The Calva Team PR Checklist:
<!-- Please read the list, since you'll get a better idea about what to expect by doing so. 😄 -->

Before merging we (at least one of us) have:

- [ ] Made sure the PR is directed at the `dev` branch (unless reasons).
- [ ] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and tested it there if so.
- [ ] Read the source changes.
- [ ] Given feedback and guidance on source changes, if needed. (Please consider noting extra nice stuff as well.)
- [ ] Tested the VSIX built from the PR (well, if this is a PR that changes the source code.)
     - [ ] Tested the particular change
     - [ ] Figured if the change might have some side effects and tested those as well.
     - [ ] Smoke tested the extension as such.
- [ ] If need be, had a chat within the team about particular changes.

Ping @pez, @kstehn, @cfehse, @bpringe

<!-- This is a nice book to read about the power of checklists: https://www.samuelthomasdavies.com/book-summaries/health-fitness/the-checklist-manifesto/ -->

PEZ commented:
Ah, great that you found where this happens! It's consistent with the behaviour that this just happens now and then.

However, we can't delete that code. It is used for the **Calva Connect** command. What you'll need to do is to only execute  it conditionally. So if we already have a port, this should not be done, otherwise it should.

PEZ commented on CHANGELOG.md:
```diff
@@ -4,6 +4,7 @@ Changes to Calva.
 
 ## [Unreleased]
 - [Use clojure-lsp for usages for debug instrumentation decorations, and stop injecting clj-kondo at jack-in](https://github.com/BetterThanTomorrow/calva/issues/931)
+- [Fix nrepl port detection race condition](https://github.com/BetterThanTomorrow/calva/issues/901)
```
```suggestion
- Fix: [nrepl port detection race condition](https://github.com/BetterThanTomorrow/calva/issues/901)
```

PEZ commented on src/connector.ts:
```diff
@@ -438,23 +438,65 @@ export let nClient: NReplClient;
 export let cljSession: NReplSession;
 export let cljsSession: NReplSession;
 
+function filePresence(portFile: vscode.Uri, timeout: number) {
```
It is unclear to me why we need a file watcher. There are mainly two paths in to the connect:

1. Jack-in. The port will be known already. No need to read the file.
2. Connect: Assumes the repl is started and and reads the file if it is there.

clyfe replied:
Removed.

clyfe commented:
Tested the VSIX built from the PR locally, but not the one off circleci, nor will I do so.

PEZ commented:
> Tested the VSIX built from the PR locally, but not the one off circleci

That's perfectly OK. The point is to make sure the release build works in the sense that the tree shaking and munging hasn't broken things. It is tested just as well with your locally built VSIX:

clyfe reviewed (commented):
Let's merge.

PEZ reviewed (changes requested):
This needs to be addressed before merge.

Commit:
Fix nrepl port race condition

Race is: the get-port-from-file code overwrites the port-from-regex with nothing when the nrepl-port file is not yet present at the time is read.
Fix: don't get-port-from-file when port already present.
Improved get-port-from-file such that it waits for file to be present with a timeout.

Fixes #901

PEZ commented on src/connector.ts:
```diff
@@ -439,22 +439,29 @@ export let cljSession: NReplSession;
 
 export async function connect(connectSequence: ReplConnectSequence,
-    isAutoConnect: boolean, hostname?: string, port?: string) {
+    isAutoConnect: boolean, maybeHostname?: string, maybePort?: string) {
```
We currently do not name optional arguments like this in the code base. It might be something we should consider, but  until we have decided on that, let's not introduce it in one place only.

clyfe replied:
Reverted.

PEZ commented on src/connector.ts:
```diff
@@ -439,22 +439,29 @@ export let cljSession: NReplSession;
-        let bytes = await vscode.workspace.fs.readFile(portFile);
-        let port = new TextDecoder("utf-8").decode(bytes);
+        let port: string | undefined = maybePort;
+        if (port === undefined) { // when we connect to a repl started externally
```
I had to think quite a while before I realised what ”externally” means here. I think we can cope without the comment. Looks clear enough to me that we are looking for a port number in the file because we don't have a port number yet. Then again, maybe the function as such lacks a documentation section which could inform about that if a port is not supplied we will search in the port file.

clyfe replied:
Removed.

PEZ commented on src/connector.ts:
```diff
@@ -439,22 +439,29 @@ export let cljSession: NReplSession;
+        let port: string | undefined = maybePort;
+        if (port === undefined) { // when we connect to a repl started externally
+            const portFile = projectTypes.nreplPortFileUri(connectSequence);
+            // Most of the time the repl started externally is of project provenance
+            // and the port is the one in the nrepl file, use it if there is one.
+            // Noting that's not always the case; if the nrepl is from another project
+            // the port guessed from the opened folder is not the one we want, and
+            // it needs to be manually changed to the right one.
```
I think the code is clear enough here. Adding the comments makes bit harder to scan the function.

clyfe replied:
Removed.

PEZ reviewed (changes requested):
Thanks. We're getting there. 😄 

Generally (and we should add this to the contribution guidelines) we want fixes like this to be quite surgical. You've found the place for the race condition, you've fixed it. If you see other changes that should be made, better send them in a separate PR.

PEZ reviewed (approved):
Thanks. Totally appreciate that you helped fix this and followed through.
